compositor_render/src/transformations/web_renderer/packet.rs

```diff
@@ -0,0 +1,56 @@
+use std::{
+    io::{Read, Write},
+    net::TcpStream,
+};
+
+pub struct Packet<'a>(pub &'a [u8]);
+
+impl<'a> Packet<'a> {
+    pub fn send(&self, stream: &mut TcpStream) -> Result<(), PacketError> {
+        let header = (self.0.len() as u32).to_be_bytes();
+        let buffer = [&header, self.0].concat();
+        stream.write(&buffer).map_err(PacketError::SendFailure)?;
+
+        Ok(())
+    }
+
+    pub fn read(stream: &mut TcpStream) -> Result<Vec<u8>, PacketError> {
```

Comment:
This function could use some rewrite:
- better naming e.g. `len` of what?  msg?
- `4 - header.len() + buf_offset` avoid arithmetic like that, approach like that removes a lot of safety of rust.
- Use a range like `[buf_offset..n]` only when it's clear the the values are correct, if some more complicated logic is involved and it's not avoidable use function that will return result and won't panic
- consider using bytes package instead of Vec

In terms of more general code structure:
- impl on Packet might be responsible for serialization, but sending/reading from socket should not be part of the implementation. If you want to keep current signature of read/send function it should be on some wrapper around TcpStream struct
- When the last read is executed it can read the end of the current packet and the start of the new one. In this implementation, you are losing that information.

let me know if you are not sure how to approach it